lightning-transaction-sync/src/electrum.rs

```diff
@@ -0,0 +1,442 @@
+							let script_history = self.client.script_get_history(&tx_out.script_pubkey)?;
+							if let Some(filtered) = script_history.iter()
+								.filter(|h| h.tx_hash == txid).max_by_key(|x| x.height)
+								{
+									let prob_conf_height = filtered.height;
+									let block_header = self.client.block_header(
+										prob_conf_height as usize)?;
+									if block_header.block_hash() == block_hash {
+										// Skip if the tx is still confirmed in the block in question.
+										continue;
+									}
+								}
+						}
+					}
+					Err(electrum_client::Error::Protocol(_)) => {
+						// We couldn't find the tx, unconfirming..
+					}
+					Err(e) => {
+						log_error!(self.logger, "Failed to look up transaction {}: {}.", txid, e);
+						return Err(InternalError::Failed);
+					}
+				}
+
+				unconfirmed_txs.push(txid);
+			} else {
+				log_error!(self.logger,
+					"Untracked confirmation of funding transaction. Please ensure none of your channels had been created with LDK prior to version 0.0.113!");
+				panic!("Untracked confirmation of funding transaction. Please ensure none of your channels had been created with LDK prior to version 0.0.113!");
+			}
+		}
+		Ok(unconfirmed_txs)
+	}
+
+	fn get_confirmed_tx(&self, tx: &Transaction, prob_conf_height: u32)
+		-> Result<ConfirmedTx, InternalError>
+	{
+		let txid = tx.txid();
+		match self.client.transaction_get_merkle(&txid, prob_conf_height as usize) {
+			Ok(merkle_res) => {
+				debug_assert_eq!(prob_conf_height, merkle_res.block_height as u32);
+				match self.client.block_header(prob_conf_height as usize) {
+					Ok(block_header) => {
+						if !self.validate_merkle_proof(&txid,
+							&block_header.merkle_root, &merkle_res)?
+						{
+							log_trace!(self.logger,
+								"Inconsistency: Block {} was unconfirmed during syncing.",
+								block_header.block_hash());
+							return Err(InternalError::Inconsistency);
+						}
+						let confirmed_tx = ConfirmedTx {
+							tx: tx.clone(),
+							block_header, block_height: prob_conf_height,
+							pos: merkle_res.pos
+						};
+						Ok(confirmed_tx)
+					}
+					Err(e) => {
+						log_error!(self.logger,
+							"Failed to retrieve block header for height {}: {}.",
+							prob_conf_height, e);
+						Err(InternalError::Failed)
+					}
+				}
+			}
+			Err(e) => {
+				log_trace!(self.logger,
+					"Inconsistency: Tx {} was unconfirmed during syncing: {}",
+					txid, e);
+				Err(InternalError::Inconsistency)
+			}
+		}
+	}
+
+	/// Returns a reference to the underlying Electrum client.
+	pub fn client(&self) -> &ElectrumClient {
+		&self.client
+	}
+
+	fn validate_merkle_proof(&self, txid: &Txid, merkle_root: &TxMerkleNode,
+		merkle_res: &GetMerkleRes) -> Result<bool, InternalError>
+	{
+		let mut index = merkle_res.pos;
+		let mut cur = txid.as_hash();
+		for bytes in &merkle_res.merkle {
+			let mut reversed = Vec::with_capacity(32);
+			reversed.truncate(0);
+			reversed.extend(bytes.iter().rev());
```

Comment:
If it's feasible to change the API to take `GetMerkleRes` by value, then you can use `reverse` on a mutable `bytes` slice to avoid allocating a new `Vec`.

Reply:
If there's push back, I'd recommend using a byte array on the stack instead of a `Vec` which results in the heap allocation.